rfcs/2024-03-22-20170-trace-data-model.md

```diff
@@ -0,0 +1,196 @@
+# RFC 20170 - 2024-03-22 - Establish an Internal Trace Model
+
+As of today there is an already existing Log and Metric event model, the purpose is to establish an
+internal model for describing a trace event.
+
+## Context
+
+The [Ingest OpenTelemetry Traces RFC] was accepted, but there was a condition that an [internal trace model] MUST be
+established. Meanwhile, there's also the [Accept Datadog Traces RFC] which was established, but
+that primarily only works well with the [`datadog_traces` sink]. There was also a previous RFC which
+was concerned with [validating schemas], but never came attempted to define an event schema.  This leaves this RFC
+to establish that data model.
+
+## Cross cutting concerns
+
+- This is part of the effort to support [sending OpenTelemetry traces], it would also support
+  ingesting OpenTelemetry. As well as to provide a path for a canonical model which traces can be
+  translated to/from ideally without loss of fidelity.
+
+## Scope
+
+### In scope
+
+- Define a Vector Trace event model schema.
+- Support for Links between `TraceEvent`s.
+- Defining a mapping between the Vector trace event model to the Datadog trace event (one direction).
+- Defining a mapping between the Vector trace event model and the OpenTelemetry trace event
+  (bi-directional).
+
+### Out of scope
+
+- Change the `TraceEvent` type internally from a re-typing of `LogEvent`.
+- Define the settings and configuration for sinks and sources using the data model.
+
+## Pain
+
+- Without an internal trace model it makes it more difficult to manage adding `sinks`, `sources`,
+  and `transforms` to Vector allowing for translations functions between the various components.
+
+## Proposal
+
+### User Experience
+
+- As a Vector user I would like to be able to ingest and propagate traces using various protocols.
+  This will allow for the greatest flexibility when managing and migrating services of developed by
+  a set of different sources, including closed and open source software.
+
+### Implementation
+
+For the initial Implementation, the data model will be [v1.1.0][otel-proto-110] of the OpenTelemetry Proto.
```

Comment:
Some general description of what is in the data model would be useful here. Also, is this data model so generic that only a `Map` can describe it, or are more dedicated data types relevant?

Reply:
I think this maybe is resolved, perhaps con be continued in other specific threads.